src/com/ms/silverking/cloud/dht/meta/DHTConfigurationWatcher.java

```diff
@@ -16,99 +16,99 @@
-    
-    public DHTConfigurationWatcher(ZooKeeperConfig zkConfig, String dhtName, long intervalMillis) 
-                  throws IOException, KeeperException {
-        this(zkConfig, dhtName, intervalMillis, true);
-    }
-    
-    public ZooKeeperConfig getZKConfig() {
-        return zkConfig;
-    }
-    
-    public MetaClient getMetaClient() {
-        return mc;
-    }
-    
-    public void waitForDHTConfiguration() {
-        synchronized (this) {
-            while (dhtConfig == null) {
-                try {
-                    this.wait();
-                } catch (InterruptedException ie) {
-                }
-            }
-        }
-    }
-    
-    public void addListener(DHTConfigurationListener listener) {
-        dhtConfigurationListeners.add(listener);
-    }
-    
-    public DHTConfiguration getDHTConfiguration() {
-        return dhtConfig;
-    }
-    
-    /**
-     * Called when a new DHT configuration is found. Reads the configuration notifies listeners.
-     */
-    private void newDHTConfiguration() {
+  private final MetaClient mc;
+  private final MetaPaths mp;
+  private final ZooKeeperConfig zkConfig;
+  private volatile DHTConfiguration dhtConfig;
+  private final boolean enableLogging;
+  private final List<DHTConfigurationListener> dhtConfigurationListeners;
+
+  public DHTConfigurationWatcher(ZooKeeperConfig zkConfig, String dhtName, long intervalMillis, boolean enableLogging)
+      throws IOException, KeeperException {
+    mc = new MetaClient(dhtName, zkConfig);
+    mp = mc.getMetaPaths();
+    this.zkConfig = zkConfig;
+    new VersionWatcher(mc, mp.getInstanceConfigPath(), this, intervalMillis, 0);
+    this.enableLogging = enableLogging;
+    dhtConfigurationListeners = Collections.synchronizedList(new ArrayList<>());
+  }
+
+  public DHTConfigurationWatcher(ZooKeeperConfig zkConfig, String dhtName, long intervalMillis)
+      throws IOException, KeeperException {
+    this(zkConfig, dhtName, intervalMillis, true);
+  }
+
+  public ZooKeeperConfig getZKConfig() {
+    return zkConfig;
+  }
+
+  public MetaClient getMetaClient() {
+    return mc;
+  }
+
+  public void waitForDHTConfiguration() {
+    synchronized (this) {
+      while (dhtConfig == null) {
         try {
-            synchronized (this) {
-                DHTConfiguration    _dhtConfig;
-                
-                _dhtConfig = mc.getDHTConfiguration();
-                dhtConfig = _dhtConfig;
-                if (enableLogging) {
-                    Log.warning("DHTConfigurationWatcher.newDHTConfiguration: "+ dhtConfig);
-                }
-                if (dhtConfig != null) {
-                    this.notifyAll();
-                } else {
-                    Log.warning("Ignoring null dhtConfig");
-                }
-                notifyListeners(_dhtConfig);
-            }
-        } catch (Exception e) {
-            Log.logErrorWarning(e);
+          this.wait();
+        } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)